### PR TITLE
Create algolia records using sitemap.xml

### PR DESCRIPTION
**Pull request by enisdenjo**
`_meta.json` has been replaced by `_meta.{ts,tsx,js,jsx}` starting Nextra v3. The _meta files now can contain executable code, components and other JS things - hence cannot be simply "parsed". On the other hand, executing the _meta files can get tricky and nasty real quick.

Instead of parsing/executing the _meta files, generating algolia records is now done through reading the `sitemap.xml`.

### CODE REVIEW

File: packages/algolia/src/bin.ts [hunk 1]
```diff
@@ -12,14 +12,16 @@ program
   .option('-o, --output [lockfilepath]', 'relative path to lock file', './algolia-lockfile.json')
   .option('-p, --publish')
   .requiredOption('-s, --source <source>')
-  .requiredOption('-d, --domain <domain>');
+  .requiredOption('-d, --domain <domain>')
+  .requiredOption('--sitemap-xml <sitemapXmlPath>');
```
Comment (B2o5T) [inline packages/algolia/src/bin.ts:16]: Do we need this option? It should be always under root/public/sitemap.xml?

File: packages/algolia/src/index.ts [hunk 2]
```diff
@@ -140,103 +136,111 @@ interface IndexToAlgoliaNextraOptions {
...
+  if (!sitemapXmlPath) {
+    throw new Error('Sitemap path not provided');
+  }
+  if (!existsSync(sitemapXmlPath)) {
+    throw new Error(`Sitemap at path "${sitemapXmlPath}" does not exist`);
+  }
+  const sitemap = new XMLParser().parse(await readFile(sitemapXmlPath, 'utf-8'));
+  const sitemapUrls: { loc: string }[] = sitemap?.urlset?.url;
+  if ((sitemapUrls?.length || 0) === 0) {
+    console.log(sitemapXmlPath, JSON.stringify(sitemap, null, '  '));
```
Comment (B2o5T) [inline packages/algolia/src/index.ts:153]:
```suggestion
```

File: packages/algolia/src/index.ts [hunk 3]
```diff
@@ -140,103 +136,111 @@ interface IndexToAlgoliaNextraOptions {
...
+    throw new Error('Sitemap urls not found at path urlset>url');
+  }
+  const urls = sitemapUrls.map(({ loc }) => loc).filter(Boolean);
+  if ((urls?.length || 0) === 0) {
+    console.log(sitemapXmlPath, JSON.stringify(sitemap, null, '  '));
```
Comment (B2o5T) [inline packages/algolia/src/index.ts:158]:
```suggestion
```

File: packages/algolia/src/index.ts [hunk 4]
```diff
@@ -140,103 +136,111 @@ interface IndexToAlgoliaNextraOptions {
...
+    throw new Error(`Sitemap at path "${sitemapXmlPath}" does not exist`);
+  }
+  const sitemap = new XMLParser().parse(await readFile(sitemapXmlPath, 'utf-8'));
+  const sitemapUrls: { loc: string }[] = sitemap?.urlset?.url;
+  if ((sitemapUrls?.length || 0) === 0) {
+    console.log(sitemapXmlPath, JSON.stringify(sitemap, null, '  '));
+    throw new Error('Sitemap urls not found at path urlset>url');
+  }
+  const urls = sitemapUrls.map(({ loc }) => loc).filter(Boolean);
+  if ((urls?.length || 0) === 0) {
```
Comment (B2o5T) [inline packages/algolia/src/index.ts:157]:
```suggestion
  if (!urls?.length) {
```

File: packages/algolia/src/index.ts [hunk 5]
```diff
@@ -140,103 +136,111 @@ interface IndexToAlgoliaNextraOptions {
...
+async function getSitemapUrlsXml(sitemapXmlPath: string): Promise<string[]> {
+  if (!sitemapXmlPath) {
+    throw new Error('Sitemap path not provided');
+  }
+  if (!existsSync(sitemapXmlPath)) {
+    throw new Error(`Sitemap at path "${sitemapXmlPath}" does not exist`);
+  }
+  const sitemap = new XMLParser().parse(await readFile(sitemapXmlPath, 'utf-8'));
+  const sitemapUrls: { loc: string }[] = sitemap?.urlset?.url;
+  if ((sitemapUrls?.length || 0) === 0) {
```
Comment (B2o5T) [inline packages/algolia/src/index.ts:152]:
```suggestion
  if (!sitemapUrls?.length) {
```

File: packages/algolia/src/bin.ts (hunk 1 above)
Comment (enisdenjo) [reply packages/algolia/src/bin.ts:16]: We can provide a default, but I'd keep it here.

File: packages/algolia/src/index.ts (hunk 2 above)
Comment (enisdenjo) [reply packages/algolia/src/index.ts:153]: Intentionally keeping it there to understand how the sitemap looks like when the command errors out.

File: packages/algolia/src/index.ts (hunk 3 above)
Comment (enisdenjo) [reply packages/algolia/src/index.ts:158]: Intentionally keeping it there to understand how the sitemap looks like when the command errors out.

File: packages/algolia/src/index.ts (hunk 2 above)
Comment (B2o5T) [reply packages/algolia/src/index.ts:153]: Let’s change to console.error instead?

File: packages/algolia/src/index.ts (hunk 3 above)
Comment (B2o5T) [reply packages/algolia/src/index.ts:158]: Console.error is better?

File: packages/algolia/src/index.ts [hunk 6]
```diff
@@ -140,103 +136,111 @@ interface IndexToAlgoliaNextraOptions {
...
+    } else if (((pagePath = `${pagePath}x`), existsSync(pagePath))) {
+      // .mdx
+      pageContent = await readFile(pagePath, 'utf-8');
+    } else {
+      throw new Error(`Page ${pagePath.replace('.mdx', '.{md,mdx}')} does not exist`);
+    }
+
+    // front-matter title or first appearing H1 heading content
+    const {
+      data: { title: matterTitle, type: matterType },
```
Comment (B2o5T) [inline packages/algolia/src/index.ts:197]: I don’t think we use frontmatter `type` somewhere in docs

File: packages/algolia/src/index.ts [hunk 7]
```diff
@@ -140,103 +136,111 @@ interface IndexToAlgoliaNextraOptions {
...
-    }
-    const [title, hierarchy, urlPath] = metaData;
 
     objects.push({
       objectID: slugger.slug(
-        `${objectsPrefix}-${[...hierarchy, filename.replace('.', '_')].join('-')}`,
+        `${objectsPrefix}-${[
+          ...hierarchy,
+          // TODO: should we remove the `index` file from the object ID?
+          filenameWithoutExt.replace('.', '_'),
```
Comment (B2o5T) [inline packages/algolia/src/index.ts:229]:
```suggestion
          filenameWithoutExt.replaceAll('.', '_'),
```

Comment (B2o5T) [reply packages/algolia/src/index.ts:229]: Since you can have more dots in filename

**Review by B2o5T — APPROVED** [review]

File: packages/algolia/src/index.ts (hunk 3 above)
Comment (enisdenjo) [reply packages/algolia/src/index.ts:158]: The log is not an error, its just an info about the contents of the xml. The error is thrown below. We could however change it to console.debug but the subtleties are not that important imho...

**Review by kamilkisiela — APPROVED** [review]

File: packages/algolia/src/index.ts (hunk 6 above)
Comment (enisdenjo) [reply packages/algolia/src/index.ts:197]: Let's leave it just in case

**Pull request merged**